### PR TITLE
Switch AppEngine Standard to Go 1.12

### LINKED ISSUE #1747: Migrate Go AppEngine to use cloud SDK and eventually to golang 1.12
## Goal

Upgrade standard AppEngine Go runtime to 1.12+ early to avoid last-minute rush before deprecation as it's a relatively big refactoring.

It also has additional benefits:
* Greatly simplify our code base (unifying flex and standard runtimes)
* Open up access to more Cloud SDK (no longer need special AppEngine versions)
* Modernize our toolchain to benefit from new Go features and bring the prod environment closer to our default dev setup on workstations (which has the latest Go)

## Overview

The biggest blocker for upgrading to 1.12+ for us is:
> The Go 1.12+ runtimes do not provide access to proprietary App Engine APIs. You can use the Google Cloud client library or third party libraries to access Google Cloud services. You can find suggested alternatives to App Engine APIs in the Migrating from the App Engine Go SDK section.

as we've done the [rest](https://cloud.google.com/appengine/docs/standard/go/go-differences#changes_in_the_app_engine_go_112_runtime) when upgrading to Go 1.11.

Here's the full list of APIs for reference:
https://cloud.google.com/appengine/docs/standard/go/go-differences#migrating-appengine-sdk

Note that we will be able to migrate gradually, **in place** -- we can keep using the Go 1.11 runtime while switching individual APIs to the new Cloud SDK, since most of them are available in Go 1.11 as well:
https://cloud.google.com/appengine/docs/standard/go111/go-differences#migrating-appengine-sdk

## Breakdown

* [x] Clean up remaining urlfetch (tiny) 
* [x] AppEngine Modules API -> google-api-go-client (small)
* [x] TaskQueue -> Cloud Tasks (small)
* [x] AppEngine logging -> Cloud Logging (small) @KyleJu 
* [x] AppEngine Datastore -> Cloud Datastore (big, need to figure out standalone Datastore emulator) (refactoring by @KyleJu, switchover by @Hexcles )
* [x] AppEngine Context -> request.Context (small) @KyleJu 
* [x] AppEngine memcache -> Memorystore (medium)
* [x] AppEngine Users API -> GitHub OAuth login (big)
* [x] Remove AppEngine search (never turned on by default) (small)
* [ ] Local testing: `dev_appserver` -> `go run` (small)

### PR DESCRIPTION
**Pull request by Hexcles**
This fixes #1747 (finally... or I hope?).

Please review by commits. The "core" part of the change is really the first and the fifth commit (the "flag day" change). The three commits in between are again fixing test setup in the new world.

### CODE REVIEW

**Comment by Hexcles** [comment]
(@KyleJu feel free to defer to Stephen)

**Comment by KyleJu** [comment]
> (@KyleJu feel free to defer to Stephen)

I will review this one :). I will get back to you by Monday

File: webapp/web/main.go [hunk 1]
```diff
@@ -33,5 +34,12 @@ func main() {
...
 	defer shared.Clients.Close()
-	appengine.Main()
+
+	// This behaves differently in prod and locally:
+	// * Prod: the provided app.yaml is not used; it simply starts the
+	//   DefaultServerMux on $PORT (defaults to 8080).
+	// * Local: in addition to the prod behaviour, it also starts some
+	//   static handlers according to app.yaml, which effectively replaces
+	//   dev_appserver.py.
+	nicehttp.Serve("webapp/web/app.staging.yaml", nil)
```
Comment (KyleJu) [inline webapp/web/main.go:44]: can we just not run it in prod (nested it under `if isDevAppserver()`)?

File: Makefile [hunk 2]
```diff
@@ -44,7 +44,10 @@ python_test: python3 tox
 go_build: git mockgen packr2
 	make webapp_node_modules_prod
 	go generate ./...
+	# Check all packages without producing any output.
 	go build ./...
+	# Build the webapp.
+	go build ./webapp/web
```
Comment (KyleJu) [inline Makefile:50]: why is this being run separately? it wasn't built prior to this change?

File: webapp/web/main.go (hunk 1 above)
Comment (Hexcles) [reply webapp/web/main.go:44]: We could. That was in fact my first implementation, but later I realized that `nicehttp.Serve` does basically exactly the same thing line by line except that the condition is flipped (it has its `IsProdAppEngine`). Do you prefer the explicitness?

File: Makefile (hunk 2 above)
Comment (Hexcles) [reply Makefile:50]: No, when given multiple packages (e.g. `./...`), `go build` does not store the output, so we'd need to have a separate invocation. This is built for local dev (as well as large/webdriver tests). Essentially this binary is the entrypoint to our server both on prod and locally now.

File: webapp/web/main.go (hunk 1 above)
Comment (KyleJu) [reply webapp/web/main.go:44]: Right I did notice the prod check in `nicehttp.Serve`, but why do they[start a localhost:8080](https://github.com/samthor/nicehttp/blob/0bb46c93be00919c226a1bf06908e455f31f62d2/gae.go#L161) in prod?

Comment (Hexcles) [reply webapp/web/main.go:44]: That's in line with https://cloud.google.com/appengine/docs/standard/go/go-differences#writing_a_main_package

Comment (KyleJu) [reply webapp/web/main.go:44]: I see

File: webdriver/webapp_server.go [hunk 3]
```diff
@@ -126,146 +104,19 @@ func NewWebserver() (s AppServer, err error) {
...
-	return nil
-}
-
 func addStaticData(i *devAppServerInstance) (err error) {
 	cmd := exec.Command(
 		"go",
 		"run",
 		"../util/populate_dev_data.go",
-		fmt.Sprintf("--project=%s", i.project),
-		fmt.Sprintf("--datastore_host=127.0.0.1:%d", i.gcdPort),
```
Comment (KyleJu) [inline webdriver/webapp_server.go:261]: Don't we still need this?

Comment (Hexcles) [reply webdriver/webapp_server.go:261]: No, `sharedtest.NewAEInstance` sets the env vars which will be picked up by `datastore.NewClient`.

https://github.com/web-platform-tests/wpt.fyi/blob/4720ce1316df824a536e0cf933c7ddfc82f64f55/shared/sharedtest/util.go#L91-L92

File: webdriver/appserver.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  }
+
+  _awaitReady(process) {
+    return new Promise(resolve => {
+      const logDevAppserver = debug('wpt.fyi:appserver');
+      process.stderr.on('data', buffer => {
+        logDevAppserver(buffer.toString());
+      });
+
+      // TODO(Hexcles): do we still need to check if the server is up?
```
Comment (KyleJu) [inline webdriver/appserver.js:56]: Reminder here in case you forget

Comment (Hexcles) [reply webdriver/appserver.js:56]: Thanks. I'll leave this TODO as is for now. The server starts almost instantaneously.

**Review by KyleJu — APPROVED** [review]
LGTM

File: webdriver/webapp_server.go (hunk 3 above)
Comment (KyleJu) [reply webdriver/webapp_server.go:261]: yea you are right

**Comment by Hexcles** [comment]
Thanks for reviewing this big change, Kyle!

**Pull request merged**